corehq/ex-submodules/casexml/apps/phone/tests/test_cleanliness.py

```diff
@@ -0,0 +1,108 @@
+                    CaseRelationship(),
+                ]
+            )
+        )
+        self.assert_owner_clean()  # this is an actual assertion
+
+    @property
+    def owner_cleanliness(self):
+        return OwnershipCleanliness.objects.get_or_create(
+            owner_id=self.owner_id,
+            domain=self.domain,
+            defaults={'is_clean': True}
+        )[0]
+
+    def assert_owner_clean(self):
+        return self.assertTrue(self.owner_cleanliness.is_clean)
+
+    def assert_owner_dirty(self):
+        return self.assertFalse(self.owner_cleanliness.is_clean)
+
+    def _set_owner(self, case_id, owner_id):
+        case = self.factory.create_or_update_case(
+            CaseStructure(case_id=case_id, attrs={'create': False, 'owner_id': owner_id})
+        )[0]
+        self.assertEqual(owner_id, case.owner_id)
+
+    def test_add_normal_case_stays_clean(self):
+        # change the owner ID of a normal case, should remain clean
+        self.factory.create_case()
+        self.assert_owner_clean()
+
+    def test_change_owner_stays_clean(self):
+        # change the owner ID of a normal case, should remain clean
+        new_owner = uuid.uuid4().hex
+        self._set_owner(self.sample_case._id, new_owner)
+        self.assert_owner_clean()
+
+    def test_change_owner_child_case_stays_clean(self):
+        # change the owner ID of a child case, should remain clean
+        new_owner = uuid.uuid4().hex
+        self._set_owner(self.child._id, new_owner)
+        self.assert_owner_clean()
+
+    def test_add_clean_parent_stays_clean(self):
+        # add a parent with the same owner, should remain clean
+        self.factory.create_or_update_case(CaseStructure(relationships=[CaseRelationship()]))
+        self.assert_owner_clean()
+
+    def test_create_dirty_makes_dirty(self):
+        # create a case and a parent case with a different owner at the same time
+        # make sure the owner becomes dirty.
+        new_owner = uuid.uuid4().hex
+        [child, parent] = self.factory.create_or_update_case(
+            CaseStructure(
+                relationships=[
+                    CaseRelationship(CaseStructure(attrs={'owner_id': new_owner}))
+                ]
+            )
+        )
+        self.assert_owner_dirty()
+        self.assertEqual(parent._id, self.owner_cleanliness.hint)
+
+    def test_add_dirty_parent_makes_dirty(self):
+        # add parent with a different owner and make sure the owner becomes dirty
+        new_owner = uuid.uuid4().hex
+        [child, parent] = self.factory.create_or_update_case(
+            CaseStructure(
+                case_id=self.sample_case._id,
+                relationships=[
+                    CaseRelationship(CaseStructure(attrs={'owner_id': new_owner}))
+                ]
+            )
+        )
+        self.assert_owner_dirty()
+        self.assertEqual(parent._id, self.owner_cleanliness.hint)
+
+    def test_change_parent_owner_makes_dirty(self):
```

Comment:
this test is currently failing, and I can't think of any way to make it work that doesn't require at least one additional couch query (to get reverse indices). if there are reverse indices then I think there will be 2 * N additional queries at a minimum for each level of parentage (so something that is a grandparent of a case would have at least 5 queries - 1 to lookup it's children, a second to check child ownership, a third to look up all grandchildren from children, a 4th to check those ownerships, and a fifth to confirm there are no great-grandchildren).

@snopoke @dannyroberts @millerdev - do you see any way around that? or do you think that's acceptable to do on every submission?

I think the only alternative might be to store a more efficient data structure to represent case trees. which could be really useful, but also could be a lot of work.